libtenzir/include/tenzir/make_byte_reader.hpp

```diff
@@ -14,76 +14,94 @@
 namespace tenzir {
 
 /// Returns a stateful function that retrieves a given number of bytes in a
-/// contiguous buffer from a generator of chunks. The last span is underful,
+/// contiguous buffer from a generator of chunks. The last chunk is underful,
 /// i.e., smaller than the number of bytes requested, and zero-sized if the
-/// input boundaries are aligned. The function returns nullopt whenever it
+/// input boundaries are aligned. The function returns nullptr whenever it
 /// merges buffers from multiple chunks. This does not indicate completion.
 inline auto make_byte_reader(generator<chunk_ptr> input) {
   input.begin(); // prime the pump
+  return [input = std::move(input), buffer = chunk::make_empty(),
+          offset = size_t{0}](size_t num_bytes) mutable -> chunk_ptr {
+    TENZIR_ASSERT(num_bytes > 0);
+    TENZIR_ASSERT(buffer);
+    // If the buffer size exactly matches what is requested, then we can just
+    // return the buffer itself.
+    if (offset == 0 and buffer->size() == num_bytes) {
+      TENZIR_ASSERT(buffer->size() - offset == num_bytes);
+      offset = buffer->size();
+      return buffer;
+    }
+    // If we have a buffer and it's non-empty, then we can just return a slice
+    // of the buffer.
+    if (buffer->size() - offset >= num_bytes) {
+      auto result = buffer->slice(offset, num_bytes);
+      offset += num_bytes;
+      TENZIR_ASSERT(result->size() == num_bytes);
+      return result;
+    }
+    // Otherwise we need to read more.
+    auto current = input.unsafe_current();
+    if (current == input.end()) {
+      if (offset == 0) {
+        offset = buffer->size();
+        TENZIR_ASSERT(buffer->size() <= num_bytes);
+        return buffer;
+      }
+      auto result = buffer->slice(offset, num_bytes);
+      offset = buffer->size();
+      TENZIR_ASSERT(result->size() <= num_bytes);
+      return result;
+    }
+    auto chunk = std::move(*current);
+    ++current;
```

Comment:
Are we allowed to modify current after std::moved?

Reply:
We don't move the iterator, but rather the value pointed to by the iterator, as indicated by the dereferencing. We can't access `*current` again before we increment the iterator, but we don't do that here.